src/storage/mod.rs

```diff
@@ -6,20 +6,47 @@ pub(crate) use self::s3::S3Backend;
 use chrono::{DateTime, Utc};
 use failure::{err_msg, Error};
 use postgres::{transaction::Transaction, Connection};
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
 use std::ffi::OsStr;
+use std::fmt;
 use std::fs;
 use std::io::Read;
 use std::path::{Path, PathBuf};
 
 const MAX_CONCURRENT_UPLOADS: usize = 1000;
 
+pub type CompressionAlgorithms = HashSet<CompressionAlgorithm>;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CompressionAlgorithm {
+    Zstd,
+}
+
+impl std::str::FromStr for CompressionAlgorithm {
+    type Err = ();
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "zstd" => Ok(CompressionAlgorithm::Zstd),
+            _ => Err(()),
+        }
+    }
+}
+
+impl fmt::Display for CompressionAlgorithm {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CompressionAlgorithm::Zstd => write!(f, "zstd"),
+        }
+    }
+}
```

Comment:
I'm not convinced this is the best way to represent the compression algorithm.

A thing I usually strive for when creating this kind of code is making sure there is only a single place where a new variant needs to be added: this way a new contributor can just add the algorithm to the enum and follow the compiler errors to implement full support for it. This advantage is not present here, as the contributor also needs to add the algorithm name in the database through a migration, which is not intuitive nor documented anywhere.

We *could* implement some code that synchronizes the `compression` table when the application starts, but I think we can implement this more cleanly. If we remove the string representation and manually set each ID in the discriminant ([with the help of a macro](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b4f6bc42c2c6caff6377b5ec9f282318)) we'd remove a lot of complexity and make it easier for new contributors to add new variants.

Reply:
The macro is not necessary, you can use `algorithm as i32` instead. Other than that I agree, storing the id would be better than storing a string.

Reply:
While that works, there is no builtin conversion from `i32` to `CompressionAlgorithm`.

Reply:
For that, is the duplication so bad? I think `impl TryFrom<i32> for CompressionAlgorithm` would work without obscuring what's happening: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=acde692246a13cddc897fa0e1927d004

Reply:
If you're worried that doesn't give a compile error if you forget to update the `match`, we could use https://crates.io/crates/num_enum.

Reply:
Yeah my worry is it doesn't give a compilation error.

Reply:
I added a test which will fail to compile if you don't add an exhaustive match for the new variant, hopefully that's close enough. I'm not sure what this should look like for S3, if we just have a number for the content-encoding it's not super descriptive, but if it's a string we're back to the same issue of needing to update the code in many places.

Reply:
You could use `strum` to generate the `Display` and `FromStr` impls (though that doesn’t help with syncing with the database).

Reply:
As another alternative I could go with 'update code in many places' and add a test that `FromStr` is exhaustive.

Reply:
I think it's fine for S3 to use the string, but I think going with a custom macro is still the best option, especially considered how many (duplicated) impls we have.

Reply:
Ok, I went with the macro-based approach for extensibility. We can always change it later if need be.